src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -533,52 +1328,71 @@ internal static bool Dispatch()
             //
             // Note that if this thread is aborted before we get a chance to request another one, the VM will
             // record a thread request on our behalf.  So we don't need to worry about getting aborted right here.
-            //
             outerWorkQueue.MarkThreadRequestSatisfied();
 
+            //
+            // The clock is ticking!  We have ThreadPoolGlobals.TP_QUANTUM milliseconds to get some work done, and then
+            // we need to return to the VM.
+            //
+            int quantumStartTime = Environment.TickCount;
+
             // Has the desire for logging changed since the last time we entered?
-            outerWorkQueue.loggingEnabled = FrameworkEventSource.Log.IsEnabled(EventLevel.Verbose, FrameworkEventSource.Keywords.ThreadPool | FrameworkEventSource.Keywords.ThreadTransfer);
+            var enabled = FrameworkEventSource.Log.IsEnabled(EventLevel.Verbose, FrameworkEventSource.Keywords.ThreadPool | FrameworkEventSource.Keywords.ThreadTransfer);
+            if (outerWorkQueue.loggingEnabled != enabled)
+            {
+                // writing shared state.
+                outerWorkQueue.loggingEnabled = enabled;
+            }
 
             //
             // Assume that we're going to need another thread if this one returns to the VM.  We'll set this to 
             // false later, but only if we're absolutely certain that the queue is empty.
             //
             bool needAnotherThread = true;
             object outerWorkItem = null;
-            try
-            {
-                //
-                // Set up our thread-local data
-                //
-                // Use operate on workQueue local to try block so it can be enregistered 
-                ThreadPoolWorkQueue workQueue = outerWorkQueue;
-                ThreadPoolWorkQueueThreadLocals tl = workQueue.GetOrCreateThreadLocals();
-                Thread currentThread = tl.currentThread;
 
-                // Start on clean ExecutionContext and SynchronizationContext
-                currentThread.ExecutionContext = null;
-                currentThread.SynchronizationContext = null;
+            // On first successfull Dequeue we ask for another thread.
+            // If the new thread that comes is sucessful, it will ask for more and so on...
+            //
+            // We will repeat the expansion request after some number of sucesses. As long as we find work to do.
+            // If a thread finds no work due to shortage or contentions, it will exit to VM and park there.
+            //
+            int dequeueCycles = 0;
```

Comment:
probably want to define this inside the `try` or the error handling will push it to stack